### ISSUE TITLE
md suffix should be automatically added for validation

### ISSUE THREAD

**Issue by koppor** [vsch/idea-multimarkdown#493]
I am using Jekyll on my site and I want to have [Semantic URLs](https://en.wikipedia.org/wiki/Semantic_URL).

For instance, I create `FileLinks.md`, but the URL is without `md` suffix: http://help.jabref.org/en/FileLinks

This confuses the validation:

![grafik](https://user-images.githubusercontent.com/1366654/30013283-fa285de2-9145-11e7-9c43-6a7e7b2ca633.png)

Is it possible to tune the validation to also check for existance of an `md` file?

**Comment by vsch** [comment]
@koppor, you can use the link map functionality to change extensions and in this case add an extension when one does not exist. See https://github.com/vsch/idea-multimarkdown/wiki/Modifying-Link-Processing 

It is a bit more involved if you have some links with extensions and some without so you will need to change the regex to exclude any links which have an extension from being modified. Otherwise, `file.ext` will be mapped to `file.ext.md` and will not resolve.

**Comment by dkellenb** [comment]
On GitHub and GitLab it's possible to clone the whole wiki, which is markdown based. There links to other pages should be done without *.md extension. At the moment every link is marked as invalid. Also every link in the wiki starts from the wiki root.

* Example 1:
  
  Page Location: /home.md
  
  Link example: `[subpage](folder/subpage)` should link to folder/subpage.md

* Example 2:

  Page Location: /folder/subpage.md
  
  Link example: `[home](home)` should link to ../home.md

By the way, this is the correct link to [Modifying-Link-Processing](../wiki/Link-Map-Settings) (was renamed to [Modifying-Link-Settings](../wiki/Link-Map-Settings).

**Comment by vsch** [comment]
@dkellenb, GitHub wikis do not have sub-directories in links for processed pages. All pages resolve as if they are in the wiki root based on their page name. 

This page in the wiki is in extensions/Admonition-Extension.md but the link https://github.com/vsch/idea-multimarkdown/wiki/Admonition-Extension resolves but https://github.com/vsch/idea-multimarkdown/wiki/extensions/Admonition-Extension tries to download extensions.dms, and if used in the wiki with a wiki link will not resolve but try to create a new page.

However, if accessing raw text of the page it needs an extension and the subdirectory.

If a wiki page is referred to by a link with an extension then it resolves to the raw text not markdown processed result:

* https://github.com/vsch/idea-multimarkdown/wiki/Admonition-Extension markdown 
* https://github.com/vsch/idea-multimarkdown/wiki/Admonition-Extension.md does not resolve tries to create a new page.
* https://github.com/vsch/idea-multimarkdown/wiki/extensions/Admonition-Extension.md raw text

It is possible that GitLab wikis use different wiki link resolution logic and the plugin will need to implement a different link resolver for it. Extensible link resolution is a major item on the TODO list.

**Can you please clarify which location for the wrong link you are referring to? Is it in the plugin help, wiki or main repository ReadMe?**

**Comment by dkellenb** [comment]
I need to clarify:

- GitHub: Does not know subfolder, you could only reference an older version
- GitLab: Supports subfolders in their wiki

Let's have a simple example with two files in the GitHub wiki repository:
- /home.md
  ```
  # Home
  [Link to subpage](subpage)
  ```
- /subpage.md
  ```
  # Subpage
  [Link to home](home)
  ```

Both links to `subpage` and `home` are marked as *unresolved link reference* in the markdown editor. But GitHub and GitLab wiki syntax requires to link other markdown files in this way.

To pick a link to your wiki, to link the https://github.com/vsch/idea-multimarkdown/wiki/Link-Map-Settings page i could also just write `[Link-Map-Settings](../wiki/Link-Map-Settings)` which results in [Link-Map-Settings](../wiki/Link-Map-Settings)

**Comment by vsch** [comment]
@dkellenb, yes the link format can be page relative or absolute. I use absolute if I tend to copy the link from the wiki or repo to other places like comments here. I agree that for purposes of cloning the links should be relative. In the plugin there is an intention to change the link format between all available formats that resolve from the containing file to the target.

However, wiki pages need to be under their own git repo, not the main repo for the plugin to handle their resolution. This requires some configuration changes in a main project. https://github.com/vsch/idea-multimarkdown/wiki/Adding-GitHub-Wiki-to-IntelliJ-Project

The reason for it is that in the file system you can put the wiki as a subdirectory of the project. However, in the real repo main files go under `repo/blob/brach/` while wiki files go under `repo/wiki/` there is a one level directory shift which link resolution has to take into account when resolving relative links between main repo and wiki pages.

I don't think the plugin currently handles a wiki being its own project. I never thought to check if the project is the wiki. All logic for link resolution assumes the wiki is a sub-directory and a sub-module of the main repo project. I have to double check the code since this was implemented long ago and no longer in memory. It might work but I doubt it.

Keep in mind that for relative links it all depends where you are writing them. For issue comments it is `../wiki` since all issues are logically under the `repo/issues` directory. 

Main repo files need `../../wiki` since the files are logically under main `repo/blob/branch` or `repo/raw/branch` directory.

Link resolution is not straight forward and has a lot of conditions and a real PITA to test. GitHub recently fixed a bug in their wiki home page image link resolution to match the rest of the wiki pages. Previously images in the home page resolved as if the home page was `wiki.md` not `wiki/Home.md`, this broke the plugin code handling home page image links.

**Comment by vsch** [comment]
Added enhancement issue for GitLab wiki support.

**Comment by Jelmerb** [comment]
> @koppor, you can use the link map functionality to change extensions and in this case add an extension when one does not exist. See https://github.com/vsch/idea-multimarkdown/wiki/Modifying-Link-Processing
> 
> It is a bit more involved if you have some links with extensions and some without so you will need to change the regex to exclude any links which have an extension from being modified. Otherwise, `file.ext` will be mapped to `file.ext.md` and will not resolve.

How can this be achieved? Do can you point me in the right direction?

**Comment by vsch** [comment]
@Jelmerb, you need to add `Link Address Element` entry to map to convert no extension link text to file ref with `.md` extension:

|       Type       |        Regex Pattern        | Replacement |
|------------------|-----------------------------|-------------|
| Text to File Ref | `^((?:.*[\\/])?[^\.\\/]+)$` | `$1.md`     |
| File Ref to Text | `^(.+?)(?:\.md)$`           | `$1`        |

![Screenshot_Add_Md_LinkMap](https://user-images.githubusercontent.com/10299086/56067550-8e207b80-5d49-11e9-955f-45dec56ade3d.png)

This will resolve links without `.md` extension to files with `.md` extension and complete files with `.md` extension to files without an extension.

**Comment by Jelmerb** [comment]
Thank you very much!
